File: python/app/dsa/arrays/TwoSum.py

# Question : https://neetcode.io/problems/two-integer-sum/question

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        m = {}

        for i, num in enumerate(nums):
            val = target - num
            if val in m:
                return [m[val], i]
            m[num] = i
        return []
    
if __name__ == "__main__":    
    sol = Solution()
    print(sol.twoSum([2,7,11,15], 9))  # Output: [0, 1]
    print(sol.twoSum([3,2,4], 6))      # Output: [1, 2]
    print(sol.twoSum([3,3], 6))        # Output: [0, 1]

File: java/src/main/dsa/arrays/TwoSum.java
package dsa.arrays;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.IntStream;

/*
* Question : https://neetcode.io/problems/two-integer-sum/question
*/

public class TwoSum {
    
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> m = new HashMap<>();

        return IntStream.range(0, nums.length)
        .filter(i -> {
            int val = target - nums[i];
            if (m.containsKey(val)) {
                return true;
            }
            m.put(nums[i], i);
            return false;
        }).mapToObj(i -> new int[]{m.get(target - nums[i]), i})
        .findFirst()
        .orElse(new int[]{});
       
    }

    public static void main(String[] args) {
        TwoSum ts = new TwoSum();
        int[] result = ts.twoSum(new int[]{2, 7, 11, 15}, 9);
        System.out.println("Indices: [" + result[0] + ", " + result[1] + "]");
    }
}
